xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
```

Comment:
I argue that the event semantics should be designed to avoid and prevent ordering issues. Having an ordered event subsystem just exposes the chaotic ordering of the overall system. Having the event system unordered forces those inherent issues to the forefront. It is better to try to sidestep a whole class of issues.

If the order of handlers must be preserved (because there's a needed use case that must be ordered), and that order must be exposed to the user to introspect and manipulate, then each handler needs a reasonable name that a user can refer to. I do not like this idea because it allows for some very messy configuration problems that could require an in-depth understanding of xonsh and the xontribs involved.

I, for one, would much rather design it to prevent the issues than hope they never occur.